# KPMG Data Analytics Project


## Task 1 - Data Quantity Assessment

In [1]:
# Importing libraries
import numpy as np 
import pandas as pd

### Read data

In [2]:
data = pd.ExcelFile(r"C:\Users\nusay\Downloads\1KPMG.xlsx")

### Read each sheet seperately

In [ ]:
# Transaction Sheet
Transactions = pd.read_excel(data, "Transactions")

# New Customer list Sheet
NewCustomerList = pd.read_excel(data, "NewCustomerList")

# Customer Demographic Sheet
CustomerDemographic = pd.read_excel(data, "CustomerDemographic")

# Customer Address Sheet
CustomerAddress = pd.read_excel(data, "CustomerAddress")

# Transactions Dataset

In [4]:
Transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [5]:
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

#### VALIDITY Issue: product_first_sold_date has the incorrect data type, it should be a datatime

### Checking for missing values:

In [6]:
for col in Transactions.columns:
    missing_percent = np.mean(Transactions[col].isnull())
    print("{} - {}%".format(col, round(missing_percent * 100)))

transaction_id - 0%
product_id - 0%
customer_id - 0%
transaction_date - 0%
online_order - 2%
order_status - 0%
brand - 1%
product_line - 1%
product_class - 1%
product_size - 1%
list_price - 0%
standard_cost - 1%
product_first_sold_date - 1%


#### COMPLETENESS Issue: 7 out of 13 columns have missing data. The percentage of missing data is low, so the advised course of action would be, if possible, to drop the rows of the columns with missing data.



### Checking for the uniqueness of the data by looking for duplicate values:

In [7]:
Transactions.duplicated().sum()

0

### Checking for the uniqueness of each column:

In [8]:
Transactions.nunique()

transaction_id             20000
product_id                   101
customer_id                 3494
transaction_date             364
online_order                   2
order_status                   2
brand                          6
product_line                   4
product_class                  3
product_size                   3
list_price                   296
standard_cost                103
product_first_sold_date      100
dtype: int64

## Exploring datetime columns

In [11]:
Transactions['transaction_date'] = Transactions['transaction_date'].astype(str).str[:4]
Transactions["transaction_date"].value_counts()

2017    20000
Name: transaction_date, dtype: int64

## Exploring other necessary columns

In [12]:
Transactions["order_status"].value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [13]:
Transactions["brand"].value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [14]:
Transactions["product_line"].value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [15]:
Transactions["product_class"].value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [16]:
Transactions["product_size"].value_counts()

medium    12990
large      3976
small      2837
Name: product_size, dtype: int64

# New Customer List Dataset

In [17]:
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.57      0.7125     0.890625     0.757031   
1                  11        0.55      0.5500     0.687500     0.584375   
2                   5        0.76      0.7600     0.760000     0.760000   
3                   1        0.86      1.0750     1.075000     1.075000   
4                   9        0.86      0.8600     1.075000     1.075000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [18]:
NewCustomerList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

#### RELEVANCY Issue: There are 5 un-named columns within this dataset that should be dropped

In [19]:
# For the sake of data qaulity assessment, these columns are dropped:
NewCustomerList.drop(["Unnamed: 16", "Unnamed: 17", "Unnamed: 18", "Unnamed: 19", "Unnamed: 20"], axis= 1, inplace = True)

### Checking for missing values:

In [20]:
for col in NewCustomerList.columns:
    missing_percent = np.mean(NewCustomerList[col].isnull())
    print("{} - {}%".format(col, round(missing_percent * 100)))

first_name - 0%
last_name - 3%
gender - 0%
past_3_years_bike_related_purchases - 0%
DOB - 2%
job_title - 11%
job_industry_category - 16%
wealth_segment - 0%
deceased_indicator - 0%
owns_car - 0%
tenure - 0%
address - 0%
postcode - 0%
state - 0%
country - 0%
property_valuation - 0%
Rank - 0%
Value - 0%


#### COMPLETENESS Issue: There are missing values in 4 columns. Columns job_title and job_industry_category have over 10% of missing data.

### Checking for the uniqueness of the data by looking for duplicate values:


In [21]:
NewCustomerList.duplicated().sum()

0

### Checking for the uniqueness of each column:


In [22]:
NewCustomerList.nunique()

first_name                              940
last_name                               961
gender                                    3
past_3_years_bike_related_purchases     100
DOB                                     958
job_title                               184
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        1
owns_car                                  2
tenure                                   23
address                                1000
postcode                                522
state                                     3
country                                   1
property_valuation                       12
Rank                                    324
Value                                   324
dtype: int64

## Exploring datetime columns 

In [23]:
NewCustomerList['DOB'] = NewCustomerList['DOB'].astype(str).str[:4]
NewCustomerList["DOB"].value_counts()

1974    29
1978    27
1973    27
1976    27
1961    26
        ..
1946     7
1989     6
1984     5
1949     4
2002     3
Name: DOB, Length: 66, dtype: int64

## Exploring necessary columns

In [24]:
NewCustomerList["gender"].value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

#### CONSISTENCY Issue: The genders are formatted as words and not letters thus U should be replaced with Unspecified for consitency. Or Female and Male are replaced by letters F and M respectively.

In [25]:
NewCustomerList["job_title"].value_counts()

Associate Professor         15
Environmental Tech          14
Software Consultant         14
Chief Design Engineer       13
Assistant Manager           12
                            ..
Accountant II                1
Programmer IV                1
Administrative Officer       1
Accounting Assistant III     1
Web Developer I              1
Name: job_title, Length: 184, dtype: int64

In [26]:
NewCustomerList["job_industry_category"].value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [27]:
NewCustomerList["wealth_segment"].value_counts()

Mass Customer        508
High Net Worth       251
Affluent Customer    241
Name: wealth_segment, dtype: int64

In [28]:
NewCustomerList["deceased_indicator"].value_counts()

N    1000
Name: deceased_indicator, dtype: int64

In [29]:
NewCustomerList["owns_car"].value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

#### CONSISTENCY Issue: columns deceased_indicator and owns_car have a different format for Yes/No. 

# Customer Demographic Dataset

In [30]:
CustomerDemographic.head() 

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [31]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

### Checking for missing values:

In [32]:
for col in CustomerDemographic.columns:
    missing_percent = np.mean(CustomerDemographic[col].isnull())
    print("{} - {}%".format(col, round(missing_percent * 100)))

customer_id - 0%
first_name - 0%
last_name - 3%
gender - 0%
past_3_years_bike_related_purchases - 0%
DOB - 2%
job_title - 13%
job_industry_category - 16%
wealth_segment - 0%
deceased_indicator - 0%
default - 8%
owns_car - 0%
tenure - 2%


#### COMPLETENESS Issue: There are missing values in 6 columns. Columns job_title and job_industry_category have over 10% of missing data.

### Checking for the uniqueness of the data by looking for duplicate values:

In [33]:
CustomerDemographic.duplicated().sum()

0

### Checking for the uniqueness of each column:

In [34]:
CustomerDemographic.nunique()

customer_id                            4000
first_name                             3139
last_name                              3725
gender                                    6
past_3_years_bike_related_purchases     100
DOB                                    3448
job_title                               195
job_industry_category                     9
wealth_segment                            3
deceased_indicator                        2
default                                  90
owns_car                                  2
tenure                                   22
dtype: int64

## Exploring datetime columns

In [35]:
CustomerDemographic['DOB'] = CustomerDemographic['DOB'].astype(str).str[:4]
CustomerDemographic["DOB"].value_counts()

1978    217
1977    204
1976    152
1974    152
1980    123
1973    122
1979    118
1986    114
1975    114
1959     94
1985     91
1995     90
1971     88
NaT      87
1964     85
1998     80
1981     79
1994     79
1969     78
1962     76
1992     74
1997     74
1989     74
1967     73
1988     72
1987     72
1957     71
1965     69
1968     69
1996     67
1966     65
1956     63
1954     61
1993     60
1961     60
1999     60
1963     59
1960     58
1955     58
1970     57
1958     55
1972     54
1990     53
1991     48
1984     43
2000     42
1982     41
1983     40
2001     34
1953     19
2002      6
1843      1
1944      1
1931      1
1935      1
1943      1
1940      1
Name: DOB, dtype: int64

#### ACCURACY Issue - There's a DOB with a birth year of 1843 and there are 87 DOBs with a birth year NaT

## Exploring necessary columns

In [36]:
CustomerDemographic["gender"].value_counts()

Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: gender, dtype: int64

#### ACCURACY Issue: There are categories that are titled incorrectly - Female is spelled as "Femal" 
#### CONSISTENCY Issue: gender categories have different format and should be altered accordingly U - Unspecified, F - Female,  and M - Male for consistency

In [37]:
CustomerDemographic["job_title"].value_counts()

Business Systems Development Analyst    45
Tax Accountant                          44
Social Worker                           44
Internal Auditor                        42
Recruiting Manager                      41
                                        ..
Database Administrator I                 4
Health Coach I                           3
Health Coach III                         3
Research Assistant III                   3
Developer I                              1
Name: job_title, Length: 195, dtype: int64

In [38]:
CustomerDemographic["job_industry_category"].value_counts()

Manufacturing         799
Financial Services    774
Health                602
Retail                358
Property              267
IT                    223
Entertainment         136
Argiculture           113
Telecommunications     72
Name: job_industry_category, dtype: int64

In [39]:
CustomerDemographic["wealth_segment"].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [40]:
CustomerDemographic["deceased_indicator"].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [41]:
CustomerDemographic["owns_car"].value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

#### CONSISTENCY Issue: columns deceased_indicator and owns_car have a different format for Yes/No

In [42]:
CustomerDemographic["default"].value_counts()

100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

#### RELEVANCY Issue: the entire column contains no valuable data

# Customer Address Dataset

In [43]:
CustomerAddress.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [44]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


### Checking for missing values:

In [45]:
for col in CustomerAddress.columns:
    missing_percent = np.mean(CustomerAddress[col].isnull())
    print("{} - {}%".format(col, round(missing_percent * 100)))

customer_id - 0%
address - 0%
postcode - 0%
state - 0%
country - 0%
property_valuation - 0%


### Checking for the uniqueness of the data by looking for duplicate values:

In [46]:
CustomerAddress.duplicated().sum()

0

### Checking the uniquness of the columns:

In [47]:
CustomerAddress.nunique()

customer_id           3999
address               3996
postcode               873
state                    5
country                  1
property_valuation      12
dtype: int64

## Exploring necessary columns

In [48]:
CustomerAddress["state"].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

#### CONSISTENCY Issue: The states have different formats and are repeated: NSW is the same as New South Wales and VIC is the same as Victoria

In [49]:
CustomerAddress["country"].value_counts()

Australia    3999
Name: country, dtype: int64